File: tools/strfile.py
#!/usr/bin/env python3
"""
strfile.py -- read `.str`: every line of text in the game, and its voice-over.

`override/<language>/mdk2.str` is where the translations live, and the GOG
release ships five of them, so no second edition is needed to find out what
localising this game touches. See below: it touches this file and nothing
else.

    0x00  u32   2003        resource type tag: .tex 2001, .mod 2002, .str 2003
    0x04  u32   2           version
    0x08  u32               entry count
    0x0c  u32   0x18        offset of the entry table, i.e. right here
    0x10  u32               offset of the character data
    0x14  u32               offset of the sound-name table

    entry, 12 bytes:  { u32 id; u32 text; u32 sound }

`text` is a byte offset into the character data and the string is a
NUL-terminated run of **16-bit units, each holding one code-page byte**.
`sound` is an offset into the sound-name table, whose entries are ASCII in
**fixed 16-byte slots**, naming a `.wav`. Either offset may be 0xFFFFFFFF:
47 entries have no text and 338 have no voice-over.

*Refuted hypothesis: the text is UTF-16LE.* It reads as UTF-16 and is not.
No unit in any of the six files known to us exceeds 0x00FF, and the English
file uses 0x91 and 0x92 -- C1 control characters in Unicode, and the curly
quotes in cp1252. French adds 0x9c, cp1252's oe. The engine widens a
code-page byte to 16 bits; it does not encode a code point. Latin-1 is the
first 256 code points, so for the five Western languages the wrong reading
gives the right answer, and only a sixth file settles it: the 1C Russian
edition stores cp1251, where 0xC4 is a capital De and UTF-16 says A-diaeresis.
That release also ships `font.tex` and `dialogfont.tex` in its `local.zip`,
which the GOG one does not -- the glyphs come with the translation, because
what a byte above 0x7f *means* is decided by the font and not by the file.

That pairing is the interesting part. The subtitle and the line that speaks it
are stored together, so `.str` is not only a UI-string table -- it is the
script of the game. Entry 1 is "Door's broken. " with the sound `jd_doors`,
and `local/` ships `jd_doors.wav` right beside this file.

Checked to the 100% rule on the GOG English data: 686 entries, every text
offset in range and every unit below 0x100, the strings tiling the character
data with no byte unaccounted for, and the 348 sound offsets landing exactly
on the 348 slots of a 5568-byte table -- 5568 = 348 x 16, and 16 bytes is the
whole remainder of the file.

**Six languages, and the sixth was needed.** The GOG release ships
`override/{english,french,german,italian,spanish}/mdk2.str` and the 1C
release a Russian one, and all six parse byte-exactly -- strings tiling the
character data with nothing left over, sound offsets landing on exactly the
slots that exist. Five of different sizes looked like enough to call the
format understood, and on the encoding they were not: see above.

They also settle what localising this game means, and here five *were* not
enough either -- the answer is this file **and the font**, which only a
release with a different alphabet shows. Of the text: **one file is the whole
of it.** 678 ids are common to all five Western tables; the
eight the English build has on top are entries with no text at all,
placeholders the translators dropped. Between 547 and 635 of the 678 strings
are genuinely translated -- "Black Hole Grenade" becomes "Grenade a Trou
Noir", "Black Hole Granate", "Granada aguj. negro".

The **voice-over is not translated with them**. The sound each line names is
the same in 672 of 678 for French and 671 for German; where they differ it is
almost always a line that has a recording in one language and none in
another, not a different recording. Only one entry in the whole set points at
a genuinely different file: Italian id 84 says `ml4z_shwang1a` where English
says `ml4z_shwang2`.

Usage:
    python3 tools/strfile.py extracted/local/mdk2.str
    python3 tools/strfile.py extracted/local/mdk2.str --json > strings.json
    python3 tools/strfile.py extracted/local/mdk2.str --grep health
"""

from __future__ import annotations

import argparse
import json
import struct
import sys
from pathlib import Path

TYPE_STR = 2003
HEADER = 24
ENTRY = 12
SLOT = 16                 # the sound-name table's fixed record
ABSENT = 0xFFFFFFFF


class StrError(ValueError):
    pass


def _text(data: bytes, start: int, codepage: str) -> str:
    """One 16-bit unit per character, each holding one code-page byte."""
    end = start
    while data[end:end + 2] != b"\0\0":
        if end >= len(data):
            raise StrError(f"string at {start:#x} is unterminated")
        end += 2
    units = data[start:end]
    if units[1::2] != bytes(len(units) // 2):
        raise StrError(f"string at {start:#x} has a unit above 0xff, so it "
                       f"is not a widened code-page byte after all")
    return units[0::2].decode(codepage)


def parse(data: bytes, codepage: str = "cp1252") -> dict:
    tag, version, count, table, text_at, sound_at = struct.unpack_from(
        "<6I", data, 0)
    if tag != TYPE_STR:
        raise StrError(f"type tag {tag}, expected {TYPE_STR}")
    if table != HEADER or text_at != HEADER + count * ENTRY:
        raise StrError(f"table at {table:#x} and data at {text_at:#x} "
                       f"do not fit {count} entries")
    if sound_at > len(data) or (len(data) - sound_at) % SLOT:
        raise StrError(f"sound table at {sound_at:#x} is not a whole "
                       f"number of {SLOT}-byte slots")
    entries = {}
    for i in range(count):
        ident, text, sound = struct.unpack_from("<3I", data, table + i * ENTRY)
        entries[ident] = {
            "text": None if text == ABSENT else
                    _text(data, text_at + text, codepage),
            "sound": None if sound == ABSENT else
                     data[sound_at + sound:sound_at + sound + SLOT]
                         .split(b"\0")[0].decode("ascii"),
            "text_offset": None if text == ABSENT else text,
            "sound_offset": None if sound == ABSENT else sound,
        }
    return {"version": version, "text_at": text_at, "sound_at": sound_at,
            "entries": entries}


def coverage(data: bytes, table: dict) -> tuple[int, int, int, int]:
    """-> (text bytes used, text bytes present, sound slots used, present)."""
    used = sum(2 * len(e["text"]) + 2          # one unit per character, + NUL
               for e in table["entries"].values() if e["text"] is not None)
    slots = {e["sound_offset"] for e in table["entries"].values()
             if e["sound_offset"] is not None}
    return (used, table["sound_at"] - table["text_at"],
            len(slots), (len(data) - table["sound_at"]) // SLOT)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description=__doc__.split("\n")[1])
    ap.add_argument("src", type=Path)
    ap.add_argument("--json", action="store_true")
    ap.add_argument("--grep", help="print only strings containing this")
    ap.add_argument("--compare", nargs="+", type=Path, metavar="STR",
                    help="compare other language tables against this one")
    ap.add_argument("--codepage", default="cp1252",
                    help="the code page the release's font draws; the five "
                         "GOG languages are cp1252, the 1C Russian cp1251")
    args = ap.parse_args(argv)

    data = args.src.read_bytes()
    table = parse(data, args.codepage)
    entries = table["entries"]

    if args.json:
        json.dump({str(k): {"text": v["text"], "sound": v["sound"]}
                   for k, v in sorted(entries.items())},
                  sys.stdout, indent=1, ensure_ascii=False)
        return 0

    if args.compare:
        others = {p.parent.name or p.stem:
                  parse(p.read_bytes(), args.codepage)["entries"]
                  for p in args.compare}
        common = set(entries)
        for t in others.values():
            common &= set(t)
        print(f"{len(common)} ids common to all "
              f"{len(others) + 1} tables; this one has {len(entries)}")
        for name, t in others.items():
            same_sound = sum(1 for k in common
                             if t[k]["sound"] == entries[k]["sound"])
            translated = sum(1 for k in common
                             if t[k]["text"] != entries[k]["text"])
            print(f"  {name:10s} {translated:4d} strings translated, "
                  f"{same_sound}/{len(common)} name the same sound")
        return 0

    needle = (args.grep or "").lower()
    for ident, e in sorted(entries.items()):
        if e["text"] is None:
            continue
        if needle and needle not in e["text"].lower():
            continue
        voice = f"  [{e['sound']}]" if e["sound"] else ""
        print(f"{ident:5d}  {e['text']!r}{voice}")

    text_used, text_have, slots_used, slots_have = coverage(data, table)
    spoken = sum(1 for e in entries.values() if e["sound"])
    print(f"{len(entries)} entries, {spoken} with a voice-over; "
          f"text {text_used}/{text_have} bytes, "
          f"sound slots {slots_used}/{slots_have}", file=sys.stderr)
    return 0 if (text_used, slots_used) == (text_have, slots_have) else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: engine/src/formats/strfile.rs
//! `.str` — the game's whole text, and the voice-over each line names.
//!
//! One file holds it: `override/<language>/mdk2.str`, with `local.zip`'s copy
//! underneath. The menu, the inventory, the hints and the subtitles all index
//! it by number -- `scripts/strings.lua` is the table of names for those
//! numbers, so `strings.continue` is an id and not a word.
//!
//! ```text
//! 0x00  u32   2003        resource type tag: .tex 2001, .mod 2002, .str 2003
//! 0x04  u32   2           version
//! 0x08  u32               entry count
//! 0x0c  u32   0x18        offset of the entry table, i.e. right here
//! 0x10  u32               offset of the character data
//! 0x14  u32               offset of the sound-name table
//!
//! entry, 12 bytes:  { u32 id; u32 text; u32 sound }
//! ```
//!
//! `text` is a byte offset into the character data and the string is a
//! NUL-terminated run of **16-bit units, each holding one code-page byte**.
//! `sound` is an offset into the sound-name table, whose entries are ASCII in
//! fixed 16-byte slots. Either may be `0xFFFFFFFF`: 47 entries have no text
//! and 338 no voice-over.
//!
//! **The text is not UTF-16**, though it reads as it. The engine widens a
//! code-page byte to 16 bits; what a byte above 0x7f *means* is decided by
//! the font that draws it, not by this file. Latin-1 is the first 256 code
//! points, so for the five Western languages the wrong reading gives the
//! right answer -- the 1C Russian edition, which stores cp1251 and ships its
//! own `font.tex` beside it, is what settles it. So a string comes out of
//! here as **bytes**, and those bytes are what
//! [`crate::render::overlay::Font`] indexes.
//!
//! `../../tools/strfile.py` is the reference and holds all six languages to
//! the 100% rule: strings tiling the character data with nothing left over,
//! and every sound offset landing on a slot that exists.

pub const TYPE_STR: u32 = 2003;
const HEADER: usize = 0x18;
const ENTRY: usize = 12;
const SLOT: usize = 16;
const NONE: u32 = 0xFFFF_FFFF;

#[derive(Debug, PartialEq)]
pub enum Error {
    NotAStringFile(u32),
    Truncated,
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Error::NotAStringFile(t) => write!(f, "type tag {t}, expected {TYPE_STR}"),
            Error::Truncated => write!(f, "the file ends inside its own table"),
        }
    }
}

#[derive(Default)]
pub struct Strings {
    /// `id -> (the code-page bytes, the sound that speaks them)`. Ids are
    /// sparse and not in order, which is why this is a map.
    entries: std::collections::BTreeMap<u32, (Vec<u8>, Option<String>)>,
}

fn u32_at(data: &[u8], at: usize) -> Result<u32, Error> {
    data.get(at..at + 4)
        .map(|b| u32::from_le_bytes([b[0], b[1], b[2], b[3]]))
        .ok_or(Error::Truncated)
}

impl Strings {
    /// A table made here rather than read, for a test that needs text and
    /// not a file.
    pub fn synthetic(entries: &[(u32, &str)]) -> Strings {
        Strings {
            entries: entries
                .iter()
                .map(|(id, text)| (*id, (text.as_bytes().to_vec(), None)))
                .collect(),
        }
    }

    pub fn parse(data: &[u8]) -> Result<Strings, Error> {
        let tag = u32_at(data, 0)?;
        if tag != TYPE_STR {
            return Err(Error::NotAStringFile(tag));
        }
        let count = u32_at(data, 8)? as usize;
        let table = u32_at(data, 0x0c)? as usize;
        let text = u32_at(data, 0x10)? as usize;
        let sounds = u32_at(data, 0x14)? as usize;
        if table != HEADER {
            return Err(Error::Truncated);
        }
        let mut entries = std::collections::BTreeMap::new();
        for i in 0..count {
            let at = table + i * ENTRY;
            let (id, t, s) = (u32_at(data, at)?, u32_at(data, at + 4)?, u32_at(data, at + 8)?);
            let mut chars = Vec::new();
            if t != NONE {
                // 16-bit units, one code-page byte each, NUL terminated
                let mut p = text + t as usize;
                loop {
                    let unit = data.get(p..p + 2).ok_or(Error::Truncated)?;
                    let unit = u16::from_le_bytes([unit[0], unit[1]]);
                    if unit == 0 {
                        break;
                    }
                    chars.push(unit as u8);
                    p += 2;
                }
            }
            let sound = if s == NONE {
                None
            } else {
                let at = sounds + s as usize;
                let slot = data.get(at..at + SLOT).ok_or(Error::Truncated)?;
                let end = slot.iter().position(|&b| b == 0).unwrap_or(SLOT);
                Some(slot[..end].iter().map(|&b| b as char).collect())
            };
            entries.insert(id, (chars, sound));
        }
        Ok(Strings { entries })
    }

    /// The line for an id, as the code-page bytes the font is indexed by.
    pub fn text(&self, id: u32) -> Option<&[u8]> {
        self.entries.get(&id).map(|(t, _)| &t[..])
    }

    /// The `.wav` that speaks that line, for the 348 that have one.
    pub fn sound(&self, id: u32) -> Option<&str> {
        self.entries.get(&id).and_then(|(_, s)| s.as_deref())
    }

    pub fn len(&self) -> usize {
        self.entries.len()
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A file built by hand, so the offsets are known rather than trusted:
    /// two entries, the second with no sound.
    fn built() -> Vec<u8> {
        let mut out = Vec::new();
        for v in [TYPE_STR, 2, 2, HEADER as u32, 0, 0] {
            out.extend_from_slice(&v.to_le_bytes());
        }
        let table = out.len();
        out.extend_from_slice(&[0u8; 2 * ENTRY]);
        let text = out.len();
        for c in b"Door's broken. " {
            out.extend_from_slice(&(*c as u16).to_le_bytes());
        }
        out.extend_from_slice(&[0, 0]);
        let second = out.len() - text;
        for c in [0xC4u8, b'a'] {
            out.extend_from_slice(&(c as u16).to_le_bytes());
        }
        out.extend_from_slice(&[0, 0]);
        let sounds = out.len();
        let mut slot = [0u8; SLOT];
        slot[..8].copy_from_slice(b"jd_doors");
        out.extend_from_slice(&slot);

        out[0x10..0x14].copy_from_slice(&(text as u32).to_le_bytes());
        out[0x14..0x18].copy_from_slice(&(sounds as u32).to_le_bytes());
        for (i, (id, t, s)) in [(1u32, 0u32, 0u32), (7, second as u32, NONE)].iter().enumerate() {
            let at = table + i * ENTRY;
            out[at..at + 4].copy_from_slice(&id.to_le_bytes());
            out[at + 4..at + 8].copy_from_slice(&t.to_le_bytes());
            out[at + 8..at + 12].copy_from_slice(&s.to_le_bytes());
        }
        out
    }

    #[test]
    fn reads_the_text_and_the_voice() {
        let s = Strings::parse(&built()).unwrap();
        assert_eq!(s.len(), 2);
        assert_eq!(s.text(1), Some(&b"Door's broken. "[..]));
        assert_eq!(s.sound(1), Some("jd_doors"));
        assert_eq!(s.sound(7), None);
    }

    /// The byte above 0x7f stays one byte: it is a code-page index and not a
    /// code point, and widening it to UTF-8 would be the refuted reading.
    #[test]
    fn a_high_byte_is_one_byte() {
        let s = Strings::parse(&built()).unwrap();
        assert_eq!(s.text(7), Some(&[0xC4u8, b'a'][..]));
    }

    #[test]
    fn a_wrong_tag_is_an_error() {
        assert_eq!(
            Strings::parse(&2001u32.to_le_bytes()).err(),
            Some(Error::NotAStringFile(2001))
        );
    }
}
